### PR TITLE
LPS 154230 7 13

### PR DESCRIPTION
**Pull request by dantewang**
- LPS-154230 Add Servlet to support browsing of the log files for companies. When clicking the log file, the file will be downloaded.
- LPS-154230 CompanyLogServlet only is registered if the property company.log.enabled=true
- LPS-154230 Reuse the permission checker
- LPS-154230 Iterate log files in all companies when the user is omni-admin, otherwise, only list log files of the user's company
- LPS-154230 SF
- LPS-154230 Push down permission checking
- LPS-154230 SF, don't use Document Library exception
- LPS-154230 SF, not used
- LPS-154230 Expose one method to return specific company's log directory.
- LPS-154230 Apply
- LPS-154230 SF, move logic to CentralizedConfiguration
- LPS-154230 Make sure _companyLogRoutingAppender populated correctly and thread safe
- LPS-154230 SF, simplify
- LPS-154230 Print message when there is no log available instead of returning a blank page.
- LPS-154230 No need to do null checking, instanceof checking will fail if null
- LPS-154230 Use Path instead to avoid dealing with path separator
- LPS-154230 Add test (Log4jConfigUtilTest is full coverage test).
- LPS-154230 baseline
- LPS-154230 More detailed message when file is not found
- LPS-154230 Display file size.
- LPS-154230 Add the function about downloading file content by range in UI.
- LPS-154230 SF
- LPS-154230 Download the log file by using input startIndex and endIndex. If the user doesn't input any values in startIndex and endIndex <input>, directly download the log file. If the user inputs “0” in startIndex and doesn't input any value in endIndex, it will download the entire log file.
- LPS-154230 Files.readAllBytes() is not intended for reading in large files. Use our API UnsyncBufferedInputStream and UnsyncByteArrayOutputStream to read/write the large files.
- LPS-154230 Add servlet integration test.
- LPS-154230 SF, extract the common logic to private method.
- LPS-154230 Project files
- LPS-154230 SF, directly assign the int value
- LPS-154230 SF, remove redundant condition as it's covered by "start >= end"
- LPS-154230 SF, directly assign the int value
- LPS-154230 SF, inline, prepare for next
- LPS-154230 SF, suggested by IntelliJ IDEA
- LPS-154230 When content length is provided, ServletResponseUtil should honor it
- LPS-154230 Send the file with start position and length
- LPS-154230 Should use long for file size & position
- LPS-154230 Proper logic to set end index
- LPS-154230 Proper filename

### CODE REVIEW

**Comment by dantewang** [comment]
@yuhai Your integration test passes on my change. Please also test it manually. By the way, I see portal error log during the integration test. If the error is expected, please use expected log annotation or LoggerTestUtil to suppress or assert it. Thanks!

**Comment by dantewang** [comment]
@yuhai If everything is OK, please clean up the commits before you resend.

**Comment by dantewang** [comment]
ci:test:relevant

**Comment by dantewang** [comment]
ci:test

File: modules/apps/portal/portal-company-log/src/main/java/com/liferay/portal/company/log/internal/servlet/CompanyLogServlet.java [hunk 1]
```diff
@@ -199,21 +198,12 @@ private void _downloadLogFile(
...
-				unsyncBufferedInputStream.read(bytes);
-
-				unsyncByteArrayOutputStream.write(bytes, start, end - start);
+			try (FileChannel fileChannel = FileChannel.open(logFile.toPath())) {
+				fileChannel.position(start);
 
 				ServletResponseUtil.sendFile(
 					httpServletRequest, httpServletResponse, logFile.getName(),
-					unsyncByteArrayOutputStream.toByteArray(),
+					Channels.newInputStream(fileChannel), end - start,
```
Comment (yuhai) [inline modules/apps/portal/portal-company-log/src/main/java/com/liferay/portal/company/log/internal/servlet/CompanyLogServlet.java:206]: @dantewang, I saw your changes when I was ready to write new changes (I didn't careful check my email). I had one rule  by using Channels.newInputStream() as inputStream, but I didn't find how to make the param as ReadableByteChannel. Learn it from your changes. Thanks.

**Comment by yuhai** [comment]
ci:test:relevant

**Comment by yuhai** [comment]
ci:test

**Comment by yuhai** [comment]
Hi @zhangyan0701, could you please help scan the latest test result? Thanks.

**Comment by zhangyan0701** [comment]
No regressions found.

**Pull request closed without merging**